### PR TITLE
add asana to social auth

### PR DESCRIPTION
**Pull request by macqueen**
also starts passing user to `get_config` because we need the user to already have an identity associated for the settings view.

still working on the sentry-plugins bit, but figured i'd get this part up

@getsentry/plugins

### CODE REVIEW

File: src/sentry/conf/server.py [hunk 1]
```diff
@@ -301,6 +301,7 @@
     'social_auth.backends.github.GithubBackend',
     'social_auth.backends.bitbucket.BitbucketBackend',
     'social_auth.backends.trello.TrelloBackend',
+    'social_auth.backends.asana.AsanaBackend',
```
Comment (dcramer) [inline src/sentry/conf/server.py]: you dont need to add this one

**Review by dcramer — APPROVED** [review]

File: src/social_auth/backends/asana.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        except ValueError:
+            return None
+
+    def auth_complete(self, *args, **kwargs):
+        """Completes loging process, must return user instance"""
+        self.process_error(self.data)
+        params = self.auth_complete_params(self.validate_state())
+        try:
+            response = requests.post(self.ACCESS_TOKEN_URL, data=params,
+                              headers=self.auth_headers()).json()
```
Comment (mattrobenolt) [inline src/social_auth/backends/asana.py]: You need to add in a `response.raise_for_status()` or it'll never error for a non-2xx.

File: src/social_auth/backends/asana.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        response.pop('data')
+        self.process_error(response)
+        return self.do_auth(response['access_token'], response=response,
+                            *args, **kwargs)
+
+    @classmethod
+    def refresh_token(cls, token):
+        params = cls.refresh_token_params(token)
+        response = requests.post(cls.ACCESS_TOKEN_URL, data=params,
+                                 headers=cls.auth_headers())
```
Comment (mattrobenolt) [inline src/social_auth/backends/asana.py]: You probably want a `response.raise_for_status()` here too if this call fails for some reason.

File: src/social_auth/backends/asana.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    SETTINGS_KEY_NAME = 'ASANA_CLIENT_ID'
+    SETTINGS_SECRET_NAME = 'ASANA_CLIENT_SECRET'
+    REDIRECT_STATE = False
+
+    def user_data(self, access_token, *args, **kwargs):
+        """Loads user data from service"""
+        headers = {'Authorization': 'Bearer %s' % access_token}
+        try:
+            return requests.get(ASANA_USER_DETAILS_URL,
+                                headers=headers).json()['data']
```
Comment (mattrobenolt) [inline src/social_auth/backends/asana.py]: `raise_for_status()`

**Review by mattrobenolt — CHANGES_REQUESTED** [review]

File: src/sentry/conf/server.py (hunk 1 above)
Comment (mattrobenolt) [inline src/sentry/conf/server.py]: @macqueen Can you remove this plz?

File: src/sentry/static/sentry/app/components/pluginConfig.jsx [hunk 5]
```diff
@@ -84,9 +84,25 @@ const PluginConfig = React.createClass({
...
+      } else {
+        authUrl += '&next=' + encodeURIComponent(document.location.pathname);
+      }
+      return (
+        <div>
+          <div className="alert alert-warning m-b-1">
+            {data.config_error}
+          </div>
+          <a className="btn btn-primary" href={authUrl}>
+            Associate Identity
```
Comment (mattrobenolt) [inline src/sentry/static/sentry/app/components/pluginConfig.jsx]: Translation string?

File: src/sentry/conf/server.py (hunk 1 above)
Comment (macqueen) [inline src/sentry/conf/server.py]: @mattrobenolt it is required. talked to cramer about it in slack, but asana won't work unless it's included here

Comment (mattrobenolt) [inline src/sentry/conf/server.py]: :heart: it.

File: src/social_auth/backends/asana.py (hunk 3 above)
Comment (macqueen) [inline src/social_auth/backends/asana.py]: done

**Review by mattrobenolt — APPROVED** [review]
:cookie:

**Pull request merged**